pico/processor/src/main/java/io/helidon/pico/processor/PostConstructPreDestroyAnnotationProcessor.java

```diff
@@ -0,0 +1,131 @@
+import java.util.Set;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+
+import io.helidon.pico.InjectionPointInfo;
+import io.helidon.pico.tools.ToolsException;
+import io.helidon.pico.tools.TypeNames;
+
+import static io.helidon.builder.processor.tools.BuilderTypeTools.findAnnotationMirror;
+import static io.helidon.pico.tools.TypeTools.createTypeNameFromElement;
+
+/**
+ * Handling for {@link jakarta.annotation.PostConstruct} and {@link jakarta.annotation.PreDestroy}.
+ */
+public class PostConstructPreDestroyAnnotationProcessor extends BaseAnnotationProcessor<Void> {
+
+    private static final Set<String> SUPPORTED_TARGETS = Set.of(TypeNames.JAKARTA_PRE_DESTROY,
+                                                                TypeNames.JAKARTA_POST_CONSTRUCT,
+                                                                TypeNames.JAVAX_PRE_DESTROY,
+                                                                TypeNames.JAVAX_POST_CONSTRUCT);
+
+    /**
+     * Service loader based constructor.
+     *
+     * @deprecated this is a Java ServiceLoader implementation and the constructor should not be used directly
+     */
+    @Deprecated
+    public PostConstructPreDestroyAnnotationProcessor() {
+    }
+
+    @Override
+    protected Set<String> annoTypes() {
+        return SUPPORTED_TARGETS;
+    }
+
+    @Override
+    protected Set<String> contraAnnotations() {
+        return Set.of(TypeNames.PICO_CONFIGURED_BY);
+    }
+
+    @Override
+    void doInner(ExecutableElement method,
+                 Void builder) {
+        if (method.getKind() == ElementKind.CONSTRUCTOR) {
+            throw new ToolsException("Invalid use of PreDestroy/PostConstruct on " + method.getEnclosingElement() + "." + method);
+        }
+
+        boolean isStatic = false;
+        InjectionPointInfo.Access access = InjectionPointInfo.Access.PACKAGE_PRIVATE;
+        Set<Modifier> modifiers = method.getModifiers();
+        if (modifiers != null) {
+            for (Modifier modifier : modifiers) {
+                if (Modifier.PUBLIC == modifier) {
+                    access = InjectionPointInfo.Access.PUBLIC;
+                } else if (Modifier.PROTECTED == modifier) {
+                    access = InjectionPointInfo.Access.PROTECTED;
+                } else if (Modifier.PRIVATE == modifier) {
+                    access = InjectionPointInfo.Access.PRIVATE;
+                } else if (Modifier.STATIC == modifier) {
+                    isStatic = true;
+                }
+            }
+        }
+
+        if (isStatic || InjectionPointInfo.Access.PRIVATE == access) {
+            throw new ToolsException("Invalid use of a private and/or static PreDestroy/PostConstruct method on "
+                                             + method.getEnclosingElement() + "." + method);
+        }
```

Comment:
IIRC Jakarta PreDestroy can be private, is Pico going to be more strict?

Reply:
(You remember right; it is part of [the contract](https://jakarta.ee/specifications/annotations/2.1/apidocs/jakarta.annotation/jakarta/annotation/predestroy).)